Sources/GRPCCore/Transport/InProcessClientTransport.swift

```diff
@@ -0,0 +1,333 @@
+    )
+
+    let waitForConnectionStream: AsyncStream<Void>? = self.state.withLockedValue { state in
+      if case .unconnected(var unconnectedState) = state {
+        let (stream, continuation) = AsyncStream<Void>.makeStream()
+        unconnectedState.pendingStreams.append(continuation)
+        state = .unconnected(unconnectedState)
+        return stream
+      }
+      return nil
+    }
+
+    if let waitForConnectionStream {
+      for await _ in waitForConnectionStream {
+        // This loop will exit either when the task is cancelled or when the
+        // client connects and this stream can be opened.
+      }
+      try Task.checkCancellation()
+    }
+
+    let streamID = try self.state.withLockedValue { state in
+      switch state {
+      case .unconnected:
+        // The state cannot be unconnected because if it was, then the above
+        // for-await loop on `pendingStream` would have not returned.
+        // The only other option is for the task to have been cancelled,
+        // and that's why we check for cancellation right after the loop.
+        fatalError("Invalid state.")
+
+      case .connected(var connectedState):
+        let streamID = connectedState.nextStreamID
+        do {
+          try connectedState.serverTransport.acceptStream(serverStream)
+          connectedState.openStreams[streamID] = (clientStream, serverStream)
+          connectedState.nextStreamID += 1
+          state = .connected(connectedState)
+        } catch let acceptStreamError as RPCError {
+          serverStream.outbound.finish(throwing: acceptStreamError)
+          clientStream.outbound.finish(throwing: acceptStreamError)
+          throw acceptStreamError
+        } catch {
+          serverStream.outbound.finish(throwing: error)
+          clientStream.outbound.finish(throwing: error)
+          throw RPCError(code: .unknown, message: "Unknown error: \(error).")
+        }
+        return streamID
+
+      case .closed:
+        let error = RPCError(
+          code: .failedPrecondition,
+          message: "The client transport is closed."
+        )
+        serverStream.outbound.finish(throwing: error)
+        clientStream.outbound.finish(throwing: error)
+        throw error
+      }
+    }
+
+    defer {
+      clientStream.outbound.finish()
+
+      let maybeEndContinuation = self.state.withLockedValue { state in
+        switch state {
+        case .unconnected:
+          // The state cannot be unconnected at this point, because if we made
+          // it this far, it's because the transport was connected.
+          // Once connected, it's impossible to transition back to unconnected,
+          // so this is an invalid state.
+          fatalError("Invalid state")
+        case .connected(var connectedState):
+          connectedState.openStreams.removeValue(forKey: streamID)
+          state = .connected(connectedState)
+        case .closed(let closedState):
+          if closedState.openStreams.count == 1 {
```

Comment:
I think I found another bug: in `close()`, if we're in the `connected` state then we always transition to the `closed` state, but we never remove streams while in the `closed` state here. If we have multiple streams open and then call `close()` then `connect(lazily:)` won't return (until cancelled).

Reply:
Good catch - the problem was we weren't removing streams at the end of `withStream` when in `closed` state.
I've added a test covering this as well.